project/DerrickPikachu/README.rst

```diff
@@ -0,0 +1,125 @@
+
+With the information about those orange nodes.
+Choose the node which has lowest f(n), and color this node as red.
+
+.. image:: img/AStar_3.png
+
+Then extend the red node, evaluate the node which can be reached by the red node and color the node as orange.
+After finishing this process, the red node has beed visited. Therefore, color this red node to be blue.
+
+Loop this process until find a path from start node to goal node.
+
+This system need the user to specify the game detail and the game heuristic.
+Users can define their game in Python or C++. 
+It would be easy to implement if the game details were defined in Python, but the search is expected to be a little bit slower.
+
+API Description
+===============
+
+User can directly use the CLI to enjoy the basic example game (or environment) to experience the A* search engine.
+
+Or if they want to try the other game, they can implement their game environment derived from the environment class interface and override the method:
+
+:math:`state state_transition(state, action)`
+
+  state_transition defines how to transfer the current state into next state by appling an action. The first parameter is the current state, the second parameter is the desired action.
+
+:math:`action[] valid_actions(state)`
+
+  valid_actions return all the valid actions on the input state.
+
+:math:`float astar_heuristic(state)`
+
+  astar_heuristic is the needed user defined function used for A* search. This function is very important and will affect how quickly will A* search find a path.
+  In different game, there would be a different heuristic functions. User can define their own heuristic function for their own game.
+
+After defining the game environment, he/she should set the environment instance and the initial state for A* search engine as following:
+
+:math:`AStarEngine.setup(environment, state)`
+
+  The setup method will help engine to get the environment instance and the initial state, those information is used for the next step.
+
+Then execute the A* search:
+
+:math:`AStarEngine.run()`
+
+  This method will start to execte A* search and find the path for user.
+
+Engineering Infrastructure
+==========================
+
+To build the A* Search Engine in c++, I would like to use CMake to automatically build my project.
+
+Version Control: Git. Because this project will be written only by myself, I will use only one branch on development, and the master branch to be released branch.
+
+Test Framework: GTest for C++, pyUnit for Python
+
+The Documentation tool would be a rst file or markdown.
+
+I will wrap a build environment (not the game enironment method above) into a docker container image, as the result, user can easily pull the image and enjoy the AStarEngine immediately.
+
+Schedule
+========
+
+* Planning phase (6 weeks from 9/19 to 10/31): prepare proposal, study A* search and create repository
+* Week 1 (10/31): build the project docker image and project prototype
```

Comment:
I would like to use container is because when I develop on different machine, I can create the same container and don't need to worry about the development environment.

Reply:
The container hurts, but it probably isn't a bad idea to experience it yourself.